File: net/pdfjet/PathOp.cs
/*
 * PathOp.cs
 *
 * Copyright (c) 2026 PDFjet Software
 * Licensed under the MIT License. See LICENSE file in the project root.
 */
using System;
using System.Collections.Generic;

namespace PDFjet.NET {
/// <summary>A single path operation: a command and its points.</summary>
internal class PathOp {
    /// <summary>The path command, for example 'M' or 'C'.</summary>
    public char cmd;

    /// <summary>The x coordinate of the original quadratic control point.</summary>
    public float x1q;   // Original quadratic control
    /// <summary>The y coordinate of the original quadratic control point.</summary>
    public float y1q;   // point coordinates

    /// <summary>The x coordinate of the first control point.</summary>
    public float x1;    // Control point x1
    /// <summary>The y coordinate of the first control point.</summary>
    public float y1;    // Control point y1
    /// <summary>The x coordinate of the second control point.</summary>
    public float x2;    // Control point x2
    /// <summary>The y coordinate of the second control point.</summary>
    public float y2;    // Control point y2
    /// <summary>The x coordinate of the point.</summary>
    public float x;     // Initial point x
    /// <summary>The y coordinate of the point.</summary>
    public float y;     // Initial point y
    /// <summary>The arguments of the command.</summary>
    public List<String> args;

    /// <summary>Creates a path operation with the specified command.</summary>
    public PathOp(char cmd) {
        this.cmd = cmd;
        this.args = new List<String>();
    }

    /// <summary>Creates a path operation with the specified command and point.</summary>
    public PathOp(char cmd, float x, float y) {
        this.cmd = cmd;
        this.x = x;
        this.y = y;
        this.args = new List<String>();
    }

    /// <summary>Sets the two control points and the end point of a cubic curve.</summary>
    public PathOp SetCubicPoints(
            float x1, float y1,
            float x2, float y2,
            float x, float y) {
        this.x1 = x1;
        this.y1 = y1;
        this.x2 = x2;
        this.y2 = y2;
        this.x = x;
        this.y = y;
        return this;
    }
}
}

File: src/pathop.go
// pathop.go
//
// Copyright (c) 2026 PDFjet Software
// Licensed under the MIT License. See LICENSE file in the project root.

package pdfjet

// svgPathOp is used to create path objects.
// The path objects may consist of lines, splines or both.
// Please see Example_02.
type svgPathOp struct {
	cmd  rune
	x1q  float32 // Original quadratic control
	y1q  float32 // point coordinates
	x1   float32 // Control point x1
	y1   float32 // Control point y1
	x2   float32 // Control point x2
	y2   float32 // Control point y2
	x    float32 // Initial point x
	y    float32 // Initial point y
	args []string
}

// newSVGPathOp creates a path operation with the specified command.
func newSVGPathOp(cmd rune) *svgPathOp {
	pathOp := new(svgPathOp)
	pathOp.cmd = cmd
	pathOp.args = make([]string, 0)
	return pathOp
}

// newSVGPathOpXY creates a path operation with the specified command and point.
func newSVGPathOpXY(cmd rune, x, y float32) *svgPathOp {
	pathOp := new(svgPathOp)
	pathOp.cmd = cmd
	pathOp.x = x
	pathOp.y = y
	pathOp.args = make([]string, 0)
	return pathOp
}

func (path *svgPathOp) setCubicPoints(x1, y1, x2, y2, x, y float32) {
	path.x1 = x1
	path.y1 = y1
	path.x2 = x2
	path.y2 = y2
	path.x = x
	path.y = y
}
